src/puppetlabs/puppetdb/command/dlo.clj

```diff
@@ -1,210 +1,182 @@
+  [registry category]
+  (let [dns "puppetlabs.puppetdb.dlo"]
+    {:messages (counters/counter registry [dns category "messages"])
+     :filesize (counters/counter registry [dns category "filesize"])}))
+
+(defn- update-metrics
+  "Updates stats to reflect the receipt of the named command."
+  [metrics command size]
+  (let [global (metrics "global")
+        cmd (metrics command)]
+    (counters/inc! (:messages global))
+    (counters/inc! (:filesize global) size)
+    (counters/inc! (:messages cmd))
+    (counters/inc! (:filesize cmd) size))
+  metrics)
+
+(defn- ensure-cmd-metrics [metrics registry cmd]
+  (if (metrics cmd)
+    metrics
+    (assoc metrics cmd (cmd-counters registry cmd))))
+
+(defn- metrics-for-dir
+  "Updates metrics to reflect the discarded commands directly inside
+  the path; does not recurse."
+  [registry path]
+  (with-open [path-stream (Files/newDirectoryStream path)]
+    (reduce (fn [metrics p]
+              ;; Assume the trailing .json here and in
+              ;; entry-cmd-err-filename below.
+              (let [name (-> p .getFileName str)]
+                (if-let [cmd (:command (and (.endsWith name ".json")
+                                            (parse-cmd-filename name)))]
+                  (update-metrics (ensure-cmd-metrics metrics registry cmd)
+                                  cmd
+                                  (Files/size p))
+                  metrics)))
+            {"global" (cmd-counters registry "global")}
+            path-stream)))
+
+(defn- write-failure-metadata
+  "Given a (possibly empty) sequence of command attempts and an exception,
+  writes a summary of the failure to out."
+  [out attempts]
+  (let [n (count attempts)]
+    (dorun (map-indexed (fn [i {:keys [exception time]}]
+                          (.format out "%sAttempt %d @ %s\n"
+                                   (into-array Object [(if (zero? i) "" "\n")
+                                                       (- n i)
+                                                       time]))
+                          (.printStackTrace exception out))
+                        attempts))))
+
+(defn- entry-cmd-data-filename
+  [entry]
+  ;; 10291-1469469689-cat-4-foo.org.json
+  (let [id (stock/entry-id entry)
+        meta (stock/entry-meta entry)]
+    (str id \- meta)))
+
+(defn- err-filename
+  [id metadata]
+  ;; 10291-1469469689-cat-4-foo.org-err.txt
+  (assert (str/ends-with? metadata ".json"))
+  (str id \- (subs metadata 0 (- (count metadata) 5)) "-err.txt"))
+
+(defn- store-failed-command-info
+  [id metadata command attempts dir]
+  ;; We want the metdata and command so we don't have to reparse, and
+  ;; we want the command because id isn't unique by itself (given
+  ;; unknown commands).
+  (let [tmp (Files/createTempFile dir
+                                  (str "tmp-err-" id \- command) ""
+                                  (make-array FileAttribute 0))]
+    ;; Leave the temp file if something goes wrong.
```

Comment:
What's the reason for the temp file exactly? i.e why not just write to the final destination?